composer/utils/checkpoint.py

```diff
@@ -275,13 +290,26 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner):
                         download_object_or_file(object_name, file_destination, self.object_store)
                         log.debug(f'Finished downloading {relative_file_path} to {file_destination}.')
             except Exception as e:
-                # PyTorch will capture any exception of this function,
-                # and dist.all_gather_objects(exception) before raising it.
-                # If that all_gather_objects fails, the exception is never visible to user.
-                # We immediately print the exception to avoid that situation.
                 log.error(f'Exception {type(e)} raised during downloading: {str(e)}')
-                raise e
-
+                download_error = True
+
+        # PyTorch will capture any exception of this function,
+        # and dist.all_gather_objects(exception) before raising it.
+        # If that all_gather_objects fails, the exception is never visible to user.
+        # We immediately kill the process and print the exception
+        download_error_tensor = dist.get_device(None).tensor_to_device(torch.tensor(1 if download_error else 0))
+        error_by_rank = dist.all_gather(download_error_tensor)
+        failed_ranks = []
+        for rank, error in enumerate(error_by_rank):
+            if error.item() > 0:
```

Comment:
```suggestion
        for rank, error in enumerate(error_by_rank.tolist()):
            if error > 0:
```

this way you only do the device->host sync 1 time, instead of WORLD_SIZE times.